internal/keppel/validation.go

```diff
@@ -6,37 +6,109 @@ package keppel
 import (
 	"fmt"
 	"net/http"
+	"regexp"
 	"strings"
 
+	"github.com/google/cel-go/cel"
+
 	"github.com/sapcc/keppel/internal/models"
 )
 
 // ValidationPolicy represents a validation policy in the API.
 type ValidationPolicy struct {
-	RequiredLabels []string `json:"required_labels,omitempty"`
+	RequiredLabels  []string `json:"required_labels,omitempty"`
+	RuleForManifest string   `json:"rule_for_manifest,omitempty"`
 }
 
 // RenderValidationPolicy builds a ValidationPolicy object out of the
 // information in the given account model.
 func RenderValidationPolicy(account models.ReducedAccount) *ValidationPolicy {
-	if account.RequiredLabels == "" {
+	if account.RuleForManifest == "" {
 		return nil
 	}
 
-	return &ValidationPolicy{
-		RequiredLabels: account.SplitRequiredLabels(),
+	policy := ValidationPolicy{
+		RuleForManifest: account.RuleForManifest,
+	}
+
+	// for backwards compatibility, show required_labels field if the CEL expression is identical
+	// to the one that we would generate from a required_labels attribute
+	re := regexp.MustCompile(`^has\(labels\.([a-zA-Z_][a-zA-Z0-9_]*)\)(\s*&&\s*has\(labels\.([a-zA-Z_][a-zA-Z0-9_]*)\))*$`)
+	if re.MatchString(account.RuleForManifest) {
+		policy.RequiredLabels = generateRequiredLabels(account.RuleForManifest)
 	}
+
+	return &policy
 }
 
 // ApplyToAccount validates this policy and stores it in the given account model.
 func (v ValidationPolicy) ApplyToAccount(account *models.Account) *RegistryV2Error {
-	for _, label := range v.RequiredLabels {
-		if strings.Contains(label, ",") {
-			err := fmt.Errorf(`invalid label name: %q`, label)
+	// for backwards compatibility, if both validation.rule_for_manifest and validation.required_labels given,
+	// accept if and only if the presented CEL expression is the same as what we would generate from the presented list of required labels
+	if v.RuleForManifest != "" && v.RequiredLabels != nil {
+		generatedRule := generateRuleForManifest(v.RequiredLabels)
+		if generatedRule != v.RuleForManifest {
+			err := fmt.Errorf(`required labels %q do not match rule for manifest %q`, v.RequiredLabels, v.RuleForManifest)
+			return AsRegistryV2Error(err).WithStatus(http.StatusUnprocessableEntity)
+		}
+	}
+	if v.RuleForManifest != "" {
+		_, ast, celErr := BuildManifestValidationAST(v.RuleForManifest)
```

Comment:
Should we cache that in some kind or LRU?

Reply:
Yes, please use a https://pkg.go.dev/github.com/hashicorp/golang-lru/v2#Cache inside of BuildManifestValidationAST. It's enough to cache positive results.